File: astrago-platform/charts/gpu-process-exporter/scripts/http_server.go
package main

import (
	"context"
	"fmt"
	"io/ioutil"
	"log"
	"net/http"
	"os"
	"os/signal"
	"strconv"
	"syscall"
	"time"
)

const (
	metricsFile = "/tmp/metrics/gpu_metrics.prom"
	defaultPort = 8080
)

type Server struct {
	port        int
	metricsFile string
	logger      *log.Logger
}

func NewServer() *Server {
	port := defaultPort
	if p := os.Getenv("METRICS_PORT"); p != "" {
		if parsed, err := strconv.Atoi(p); err == nil {
			port = parsed
		}
	}

	return &Server{
		port:        port,
		metricsFile: metricsFile,
		logger:      log.New(os.Stdout, "[HTTP-SERVER] ", log.LstdFlags),
	}
}

func (s *Server) metricsHandler(w http.ResponseWriter, r *http.Request) {
	s.logger.Printf("GET %s from %s", r.URL.Path, r.RemoteAddr)

	// Check if metrics file exists
	if _, err := os.Stat(s.metricsFile); os.IsNotExist(err) {
		s.logger.Printf("WARNING: Metrics file not found: %s", s.metricsFile)
		w.Header().Set("Content-Type", "text/plain; version=0.0.4; charset=utf-8")
		w.WriteHeader(http.StatusOK)
		fmt.Fprintf(w, "# Metrics file not ready yet\n")
		return
	}

	// Read metrics file
	content, err := ioutil.ReadFile(s.metricsFile)
	if err != nil {
		s.logger.Printf("ERROR: Failed to read metrics file: %v", err)
		w.Header().Set("Content-Type", "text/plain")
		w.WriteHeader(http.StatusInternalServerError)
		fmt.Fprintf(w, "# Error reading metrics: %v\n", err)
		return
	}

	// Return metrics content
	w.Header().Set("Content-Type", "text/plain; version=0.0.4; charset=utf-8")
	w.Header().Set("Content-Length", strconv.Itoa(len(content)))
	w.WriteHeader(http.StatusOK)
	w.Write(content)

	s.logger.Printf("Metrics served successfully (%d bytes)", len(content))
}

func (s *Server) healthHandler(w http.ResponseWriter, r *http.Request) {
	s.logger.Printf("GET %s from %s", r.URL.Path, r.RemoteAddr)

	var status string
	var code int

	// Check if metrics file exists and is recent (within 2 minutes)
	if stat, err := os.Stat(s.metricsFile); err == nil {
		fileAge := time.Since(stat.ModTime())
		if fileAge < 2*time.Minute {
			status = "OK"
			code = http.StatusOK
		} else {
			status = fmt.Sprintf("STALE - metrics file too old (%.0f seconds)", fileAge.Seconds())
			code = http.StatusServiceUnavailable
		}
	} else {
		status = "ERROR - metrics file not found"
		code = http.StatusServiceUnavailable
	}

	w.Header().Set("Content-Type", "text/plain")
	w.Header().Set("Content-Length", strconv.Itoa(len(status)))
	w.WriteHeader(code)
	fmt.Fprint(w, status)

	s.logger.Printf("Health check: %s (code: %d)", status, code)
}

func (s *Server) rootHandler(w http.ResponseWriter, r *http.Request) {
	// Root path redirects to /metrics
	if r.URL.Path == "/" {
		s.metricsHandler(w, r)
		return
	}

	// 404 for unknown paths
	s.logger.Printf("404 - GET %s from %s", r.URL.Path, r.RemoteAddr)
	w.WriteHeader(http.StatusNotFound)
	fmt.Fprintf(w, "Not Found")
}

func (s *Server) loggingMiddleware(next http.Handler) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		start := time.Now()
		next.ServeHTTP(w, r)
		duration := time.Since(start)
		s.logger.Printf("Request completed in %v", duration)
	})
}

func (s *Server) Run() error {
	// Create metrics directory if it doesn't exist
	if err := os.MkdirAll("/tmp/metrics", 0755); err != nil {
		return fmt.Errorf("failed to create metrics directory: %v", err)
	}

	// Setup HTTP routes
	mux := http.NewServeMux()
	mux.HandleFunc("/metrics", s.metricsHandler)
	mux.HandleFunc("/health", s.healthHandler)
	mux.HandleFunc("/", s.rootHandler)

	// Create HTTP server
	server := &http.Server{
		Addr:           fmt.Sprintf(":%d", s.port),
		Handler:        s.loggingMiddleware(mux),
		ReadTimeout:    10 * time.Second,
		WriteTimeout:   10 * time.Second,
		IdleTimeout:    30 * time.Second,
		MaxHeaderBytes: 1 << 20, // 1MB
	}

	// Start server in goroutine
	go func() {
		s.logger.Printf("GPU Process Exporter HTTP Server starting on port %d", s.port)
		s.logger.Printf("Metrics endpoint: http://0.0.0.0:%d/metrics", s.port)
		s.logger.Printf("Health endpoint: http://0.0.0.0:%d/health", s.port)

		if err := server.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			s.logger.Fatalf("Failed to start HTTP server: %v", err)
		}
	}()

	// Wait for interrupt signal for graceful shutdown
	quit := make(chan os.Signal, 1)
	signal.Notify(quit, syscall.SIGINT, syscall.SIGTERM)
	<-quit

	s.logger.Println("Received termination signal, shutting down gracefully...")

	// Graceful shutdown with timeout
	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
	defer cancel()

	if err := server.Shutdown(ctx); err != nil {
		s.logger.Printf("Server forced to shutdown: %v", err)
		return err
	}

	s.logger.Println("Server shutdown completed")
	return nil
}

func main() {
	// Set up logger
	log.SetFlags(log.LstdFlags)
	log.SetPrefix("[MAIN] ")

	// Print startup info
	nodeName := os.Getenv("NODE_NAME")
	if nodeName == "" {
		nodeName = "unknown"
	}
	
	log.Printf("=== GPU Process Exporter HTTP Server (Go) ===")
	log.Printf("Node: %s", nodeName)
	log.Printf("Go version: %s", "1.19+")

	// Create and run server
	server := NewServer()
	if err := server.Run(); err != nil {
		log.Fatalf("Server error: %v", err)
	}
} 

File: astrago-platform/charts/gpu-process-exporter/scripts/http_server.py
#!/usr/bin/env python3
"""
GPU Process Exporter HTTP Server
파일 기반 GPU 메트릭을 HTTP API로 제공
"""

import http.server
import socketserver
import os
import logging
import signal
import sys
from urllib.parse import urlparse

# 로깅 설정
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s'
)
logger = logging.getLogger(__name__)

class MetricsHandler(http.server.BaseHTTPRequestHandler):
    """GPU 메트릭을 서빙하는 HTTP 핸들러"""
    
    def __init__(self, *args, **kwargs):
        self.metrics_file = '/tmp/metrics/gpu_metrics.prom'
        super().__init__(*args, **kwargs)
    
    def log_message(self, format, *args):
        """로그 메시지를 logger로 전달"""
        logger.info(f"{self.address_string()} - {format % args}")
    
    def do_GET(self):
        """GET 요청 처리"""
        try:
            parsed_path = urlparse(self.path)
            path = parsed_path.path
            
            if path == '/metrics':
                self.handle_metrics()
            elif path == '/health':
                self.handle_health()
            elif path == '/':
                self.handle_metrics()  # 기본 경로는 메트릭으로
            else:
                self.send_error(404, "Not Found")
                
        except Exception as e:
            logger.error(f"Error handling request {self.path}: {e}")
            self.send_error(500, "Internal Server Error")
    
    def handle_metrics(self):
        """GPU 메트릭 파일을 읽어서 응답"""
        try:
            if not os.path.exists(self.metrics_file):
                logger.warning(f"Metrics file not found: {self.metrics_file}")
                content = "# Metrics file not ready yet\n"
            else:
                with open(self.metrics_file, 'r', encoding='utf-8') as f:
                    content = f.read()
                
                if not content.strip():
                    content = "# Empty metrics file\n"
            
            # Prometheus 형식으로 응답
            self.send_response(200)
            self.send_header('Content-Type', 'text/plain; version=0.0.4; charset=utf-8')
            self.send_header('Content-Length', str(len(content.encode('utf-8'))))
            self.send_header('Connection', 'close')
            self.end_headers()
            
            self.wfile.write(content.encode('utf-8'))
            logger.info("Metrics served successfully")
            
        except Exception as e:
            logger.error(f"Error serving metrics: {e}")
            error_content = f"# Error reading metrics: {e}\n"
            self.send_response(500)
            self.send_header('Content-Type', 'text/plain')
            self.send_header('Content-Length', str(len(error_content.encode('utf-8'))))
            self.end_headers()
            self.wfile.write(error_content.encode('utf-8'))
    
    def handle_health(self):
        """헬스체크 엔드포인트"""
        try:
            # 메트릭 파일이 최근 2분 내에 업데이트되었는지 확인
            if os.path.exists(self.metrics_file):
                file_age = os.path.getmtime(self.metrics_file)
                import time
                if time.time() - file_age < 120:  # 2분 = 120초
                    status = "OK"
                    code = 200
                else:
                    status = "STALE - metrics file too old"
                    code = 503
            else:
                status = "ERROR - metrics file not found"
                code = 503
            
            self.send_response(code)
            self.send_header('Content-Type', 'text/plain')
            self.send_header('Content-Length', str(len(status.encode('utf-8'))))
            self.send_header('Connection', 'close')
            self.end_headers()
            
            self.wfile.write(status.encode('utf-8'))
            logger.info(f"Health check: {status} (code: {code})")
            
        except Exception as e:
            logger.error(f"Error in health check: {e}")
            error_msg = f"ERROR: {e}"
            self.send_response(500)
            self.send_header('Content-Type', 'text/plain')
            self.send_header('Content-Length', str(len(error_msg.encode('utf-8'))))
            self.end_headers()
            self.wfile.write(error_msg.encode('utf-8'))

class ThreadedTCPServer(socketserver.ThreadingMixIn, socketserver.TCPServer):
    """멀티스레드 TCP 서버"""
    allow_reuse_address = True
    daemon_threads = True

def signal_handler(signum, frame):
    """종료 시그널 핸들러"""
    logger.info(f"Received signal {signum}, shutting down gracefully...")
    sys.exit(0)

def main():
    """메인 함수"""
    PORT = int(os.environ.get('METRICS_PORT', 8080))
    
    # 시그널 핸들러 등록
    signal.signal(signal.SIGTERM, signal_handler)
    signal.signal(signal.SIGINT, signal_handler)
    
    try:
        # HTTP 서버 시작
        with ThreadedTCPServer(("", PORT), MetricsHandler) as httpd:
            logger.info(f"GPU Process Exporter HTTP Server starting on port {PORT}")
            logger.info(f"Metrics endpoint: http://0.0.0.0:{PORT}/metrics")
            logger.info(f"Health endpoint: http://0.0.0.0:{PORT}/health")
            
            # 메트릭 디렉토리 확인
            metrics_dir = os.path.dirname('/tmp/metrics/gpu_metrics.prom')
            if not os.path.exists(metrics_dir):
                os.makedirs(metrics_dir)
                logger.info(f"Created metrics directory: {metrics_dir}")
            
            httpd.serve_forever()
            
    except Exception as e:
        logger.error(f"Failed to start HTTP server: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main() 